src/interfaces/cloud.rs

```diff
@@ -0,0 +1,410 @@
+            fetch_entries_key: Some(fetch_entries_key),
+            fetch_chains_key: Some(fetch_chains_key),
+            upsert_entries_key: Some(upsert_entries_key),
+            insert_chains_key: Some(insert_chains_key),
+        })
+    }
+
+    pub fn reduce_permissions(&mut self, search: bool, index: bool) -> Result<(), FindexErr> {
+        self.fetch_entries_key =
+            reduce_option("fetch entries", &self.fetch_entries_key, search || index)?;
+        self.fetch_chains_key = reduce_option("fetch chains", &self.fetch_chains_key, search)?;
+        self.upsert_entries_key = reduce_option("upsert entries", &self.upsert_entries_key, index)?;
+        self.insert_chains_key = reduce_option("insert chains", &self.insert_chains_key, index)?;
+
+        Ok(())
+    }
+}
+
+/// If we have the permission and want to keep it, do nothing.
+/// If we don't have the permission and want to keep it, fail.
+/// If we don't want to keep it, return none.
+fn reduce_option(
+    debug_info: &str,
+    permission: &Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>,
+    keep: bool,
+) -> Result<Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>>, FindexErr> {
+    match (permission, keep) {
+        (_, false) => Ok(None),
+
+        (Some(permission), true) => Ok(Some(permission.clone())),
+        (None, true) => Err(FindexErr::Other(format!(
+            "The token provided doesn't have the permission to {debug_info}"
+        ))),
+    }
+}
+
+enum Callback {
+    FetchEntries,
+    FetchChains,
+    UpsertEntries,
+    InsertChains,
+}
+
+impl Callback {
+    pub fn get_uri(&self) -> &'static str {
+        match self {
+            Callback::FetchEntries => "fetch_entries",
+            Callback::FetchChains => "fetch_chains",
+            Callback::UpsertEntries => "upsert_entries",
+            Callback::InsertChains => "insert_chains",
+        }
+    }
+
+    pub fn get_key<'a>(
+        &self,
+        token: &'a Token,
+    ) -> &'a Option<KeyingMaterial<SIGNATURE_KEY_LENGTH>> {
+        match self {
+            Callback::FetchEntries => &token.fetch_entries_key,
+            Callback::FetchChains => &token.fetch_chains_key,
+            Callback::UpsertEntries => &token.upsert_entries_key,
+            Callback::InsertChains => &token.insert_chains_key,
+        }
+    }
+}
+
+impl FindexCloud {
+    pub fn new(token: String, base_url: Option<String>) -> Result<Self, FindexErr> {
+        Ok(FindexCloud {
+            token: Token::from_str(&token)?,
+            base_url,
+        })
+    }
+
+    async fn post(&self, callback: Callback, bytes: &[u8]) -> Result<Vec<u8>, FindexErr> {
+        let endpoint = callback.get_uri();
+
+        let key: KmacKey = callback
+            .get_key(&self.token)
+            .as_ref()
+            .ok_or(FindexErr::Other(format!(
+                "your key doesn't have the permission to call {endpoint}"
+            )))?
+            .derive_kmac_key(self.token.index_id.as_bytes());
```

Comment:
See my previous comment: it may be better if this is abstracted away by the `get_key()` method.

Reply:
Yes good idea, done in d47b268